### PR TITLE
feat(logging): implement logging service

### LINKED ISSUE #91: Logging service: Connect to message broker (listen for events)

### LINKED ISSUE #92: Logging service: Implement message handler logic

### LINKED ISSUE #93: Logging service: Implement ingestion logic into LogDNA (IBM Log Analysis)

### LINKED ISSUE #94: Logging service: Unit tests

### PR DESCRIPTION
**Pull request by jharvey10**
This closes out the remainder of the v0 logging epic (along with #192)

Closes #91, Closes #92, Closes #93, Closes #94

Full implementation of the logging service to ingest logs and shuttle them to LogDNA

### CODE REVIEW

File: services/logging/src/test/logging.controller.test.ts
Comment (francinelucca) [inline services/logging/src/test/logging.controller.test.ts:51]: why do we need to cast to any?

File: services/logging/src/test/logging.controller.test.ts
Comment (jdharvey-ibm) [reply services/logging/src/test/logging.controller.test.ts:36]: That would definitely be more clear as far as the intent of that part of the test, but unfortunately since `nestLogger` is marked as read-only in `PlatformController`, you can't set it when the object is asserted as a `PlatformController`

File: services/logging/src/test/logging.controller.test.ts
Comment (jdharvey-ibm) [reply services/logging/src/test/logging.controller.test.ts:51]: None of the fields on `data` are marked as optional (and they're all required in the map passed to `logLogged()`, so to get around this and fake an incoming request that is missing one of the fields, I first assert it as `any`, which gives unmitigated access to the object properties, delete one of the required fields and the pass it in.

As an alternative to this, I could instead have two interfaces. One called `LogLoggedMessage` and another called like `UnvalidatedLogLoggedMessage`, and only begin using `LogLoggedMessage` once validation has been performed successfully.

What are your thoughts on that as a general paradigm?

File: services/logging/src/main/decorators/Trace.ts
Comment (francinelucca) [reply services/logging/src/main/decorators/Trace.ts:18]: I don't see an `@trace` usage anywhere, is nestjs using this implicitly?

Comment (jdharvey-ibm) [reply services/logging/src/main/decorators/Trace.ts:18]: No, nobody is using it at the moment. It's something I know we'll want to use in other services, so I wanted to keep the code around since it took a bit of research to figure out how to get it working properly, but at the moment it's not being used anywhere. Eventually we should move it to a common place along with the PlatformController, Validate, and PlatformMicroservice.

File: services/logging/src/test/logging.controller.test.ts
Comment (francinelucca) [reply services/logging/src/test/logging.controller.test.ts:51]: oh ok I understand now. I don't think having the two interfaces is necessary we can leave like this unless you really want to change it

**Review by francinelucca — DISMISSED** [review]

**Review by mattrosno — DISMISSED** [review]
Only read through the markdown changes and looked for web app side effects. Found none.

File: services/logging/src/test/logging.controller.test.ts
Comment (jdharvey-ibm) [reply services/logging/src/test/logging.controller.test.ts:51]: I think I am gonna do two actually haha. I think this would be a good case for using type assertions on methods like you and I chatted about the other day. So the validator function would become like:

```ts
function validate(data: UnvalidatedMessage): data is LogLoggedMessage {
  // do all the checks to ensure data is a valid LogLoggedMessage
  // and `throw` otherwise
}
```

I just need to wait until #192 is merged so I can have access to LogLoggedMessage on this branch.

**Review by francinelucca — DISMISSED** [review]
lgtm

**Review by francinelucca — APPROVED** [review]

**Pull request merged**
